### PR TITLE
[DEV-2023] Consolidate template variables for OTP and forgot password

### PR DESCRIPTION
**Pull request by chankiyu22**
refs DEV-2023

### CODE REVIEW

File: pkg/lib/translation/template.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
+package translation
+
+import (
+	htmltemplate "html/template"
+)
+
+type EscapedString string
```
Comment (louischan-oursky) [inline pkg/lib/translation/template.go:7]: Remove this.

File: pkg/lib/authn/otp/sender.go [hunk 2]
```diff
@@ -24,9 +25,10 @@ type EndpointsProvider interface {
 }
 
 type TranslationService interface {
-	EmailMessageData(msg *translation.MessageSpec, args interface{}) (*translation.EmailMessageData, error)
-	SMSMessageData(msg *translation.MessageSpec, args interface{}) (*translation.SMSMessageData, error)
-	WhatsappMessageData(language string, msg *translation.MessageSpec, args interface{}) (*translation.WhatsappMessageData, error)
+	GetAppName() (appName string, err error)
```
Comment (louischan-oursky) [inline pkg/lib/authn/otp/sender.go:28]: Remove this. The caller should not be asked to do more things.

File: pkg/lib/authn/otp/sender.go [hunk 3]
```diff
@@ -63,13 +65,16 @@ func (m *PreparedMessage) Close() {
 }
 
 type MessageSender struct {
+	Context         context.Context
 	Translation     TranslationService
 	Endpoints       EndpointsProvider
 	Sender          Sender
 	WhatsappService WhatsappService
+	StaticAssets    translation.StaticAssetResolver
```
Comment (louischan-oursky) [inline pkg/lib/authn/otp/sender.go:73]: Remove this.

File: pkg/lib/authn/otp/sender.go [hunk 4]
```diff
@@ -63,13 +65,16 @@ func (m *PreparedMessage) Close() {
...
 }
 
-func (s *MessageSender) setupTemplateContext(msg *PreparedMessage, opts SendOptions) (any, error) {
+func (s *MessageSender) setupTemplateContext(msg *PreparedMessage, opts SendOptions) (*translation.TemplateVariables, error) {
+	uiParam := uiparam.GetUIParam(s.Context)
```
Comment (louischan-oursky) [inline pkg/lib/authn/otp/sender.go:77]: Remove this.

File: pkg/lib/authn/otp/sender.go [hunk 5]
```diff
@@ -108,20 +113,25 @@ func (s *MessageSender) setupTemplateContext(msg *PreparedMessage, opts SendOpti
...
-		Host:  s.Endpoints.Origin().Host,
-	})
-	if opts.AdditionalContext != nil {
-		template.Embed(ctx, opts.AdditionalContext)
-	}
-
-	return ctx, nil
+	appName, _ := s.Translation.GetAppName()
+
+	return translation.MakeOTPTemplateVariables(
```
Comment (louischan-oursky) [inline pkg/lib/authn/otp/sender.go:118]: You do not call this function here.

File: pkg/lib/feature/forgotpassword/service.go [hunk 6]
```diff
@@ -225,8 +220,8 @@ func (s *Service) sendEmail(email string, userID string, options *CodeOptions) e
 	}
 
 	err = s.OTPSender.Send(msg, otp.SendOptions{
-		OTP:               code,
-		AdditionalContext: ctx,
+		OTP:         code,
+		HasPassword: len(ais) > 0,
```
Comment (louischan-oursky) [inline pkg/lib/feature/forgotpassword/service.go:224]: PartialTemplateVariables

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
1. You do not change the dependencies of the caller.
2. The caller uses a new strong type.
3. In pkg/lib/translation/service.go, no more map[string]interface{}
4. Preserve the original behavior.
5. InputTemplateVariables -> PreparedTemplateVariables
6. In your interface, you take InputTemplateVariables
7. In pkg/lib/translation/service.go, you make InputTemplateVariables into PreparedTemplateVariables

File: pkg/lib/translation/template.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
+package translation
```
Comment (louischan-oursky) [inline pkg/lib/translation/template.go:1]: variables.go

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by chankiyu22** [comment]
@louischan-oursky Updated.

File: pkg/lib/translation/service.go [hunk 8]
```diff
@@ -85,38 +81,36 @@ func (s *Service) GetSenderForTestEmail() (sender string, err error) {
...
+func (s *Service) emailMessageHeader(name string, variables *PreparedTemplateVariables) (sender, replyTo, subject string, err error) {
 	t, err := s.translationMap()
 	if err != nil {
 		return
 	}
 
-	data := make(map[string]interface{})
-	template.Embed(data, args)
-	err = s.appMetadata(data)
+	err = s.appMetadata(variables)
```
Comment (louischan-oursky) [inline pkg/lib/translation/service.go:90]: This changes the original behavior.

The original behavior is 

```
data := {}
err := s.appMetadata(data)
// data = {"AppName": "blahblahblah" }

Then data is used to render sender, replyTo, and subject
```

File: pkg/lib/translation/service.go [hunk 9]
```diff
@@ -170,22 +170,20 @@ func (s *Service) EmailMessageData(msg *MessageSpec, args interface{}) (*EmailMe
...
+func (s *Service) smsMessageHeader(name string, variables *PreparedTemplateVariables) (sender string, err error) {
 	t, err := s.translationMap()
 	if err != nil {
 		return
 	}
 
-	data := make(map[string]interface{})
-	template.Embed(data, args)
-	err = s.appMetadata(data)
+	err = s.appMetadata(variables)
```
Comment (louischan-oursky) [inline pkg/lib/translation/service.go:179]: Same here.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/translation/service.go [hunk 10]
```diff
@@ -220,9 +217,14 @@ func (s *Service) SMSMessageData(msg *MessageSpec, args interface{}) (*SMSMessag
...
-	data := map[string]interface{}{}
-	template.Embed(data, args)
+func (s *Service) WhatsappMessageData(language string, msg *MessageSpec, variables *PartialTemplateVariables) (*WhatsappMessageData, error) {
+	uiParam := uiparam.GetUIParam(s.Context)
+	data := PrepareFromPartialTemplateVariables(variables, &UIParams{
+		ClientID:  uiParam.ClientID,
+		State:     uiParam.State,
+		XState:    uiParam.XState,
+		UILocales: uiParam.UILocales,
+	}, s.StaticAssets.StaticAssetURL)
```
Comment (louischan-oursky) [inline pkg/lib/translation/service.go:227]: Please add new tests for EmailMessageData, SMSMessageData, and WhatsappMessageData to show the public API does not change.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/translation/variables.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+	Code  string
+	URL   string
+	Link  string
+	Host  string
+
+	// OTP Additional context
+	HasPassword bool
+
+	// Forgot password
+	AppName  string
```
Comment (chankiyu22) [inline pkg/lib/translation/variables.go:23]: Remove this as the body should have no use.

File: pkg/lib/translation/variables.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	UILocales      string
+	URL            string
+	XState         string
+}
+
+func PrepareFromPartialTemplateVariables(
+	v *PartialTemplateVariables,
+	uiParams *UIParams,
+	staticAssetUrl func(id string) (url string, err error),
+) *PreparedTemplateVariables {
```
Comment (chankiyu22) [inline pkg/lib/translation/variables.go:49]: Move it under translation service
- Also prepare AppName

```golang
func (s *Service) PrepareTemplateVariables(variables *PartialTemplateVariables) {
    appName, err := ...
    return &PreparedTemplateVariables{
        AppName: appName,
        ...,
        StaticAssetURL: s.StaticAssets.StaticAssetURL,
    }, err
}
```

**Review by chankiyu22 — COMMENTED** [review]
@louischan-oursky Can take a look now. Note that there are some behaviour change :adore2x:

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**